src/cdk/testing/component-harness.ts

```diff
@@ -563,13 +563,38 @@ function _valueAsString(value: unknown) {
   }
   // `JSON.stringify` doesn't handle RegExp properly, so we need a custom replacer.
   try {
-    return JSON.stringify(value, (_, v) => {
+    const stringifiedValue = JSON.stringify(value, (_, v) => {
+      // JSON.stringify automatically wraps string values in `"` characters, so if we just return
+      // `v.toString()` for a regex it will wind up with extra quotes around it.
+      // e.g. `"/stuff/i"` instead of `/stuff/i`. Therefore to distinguish between regexes which
+      // don't need the quotes and strings that do, we use some custom escaping.
+      //
+      // The escaping rules are:
+      // 1. Any `#` within a regex or string becomes `@#@`.
+      // 2. Any regex is wrapped with `##`.
+      // 3. Any `"` in a regex becomes `##`.
+      //
+      // This guarantees that if we see `##`, we know it's part of a regex.
+      // e.g. the regex /stuff/i would become `"##/stuff/i##"`, while the string `##stuff##`
+      // would become `"@#@@#@stuff@#@@#@"` in the final stringified output.
+      //
+      // Knowing this, we can delete instances of `"##` and `##"` in the final output and then
+      // transform `@#@` back to `#`, giving us the string representation we want.
       if (v instanceof RegExp) {
-        return `/${v.toString()}/`;
+        // Replace all `#` in the regex with an escaped `@#@` and wrap it in extra `##`s.
+        return `##${v.toString().replace(/#/g, '@#@').replace(/"/g, '##')}##`;
```

Comment:
Couldn't we simplify the escaping and unescaping by wrapping the stringified regex values in something super specific that has a really low chance of collision? We do [something similar in the table data source](https://github.com/angular/components/blob/master/src/material/table/table-data-source.ts#L229). Part of the problem here is that `@` and `#` are somewhat used characters so we have to escape them in the string too. With something like `◬ngEscapeRegex◬/foo/g◬ngEscapeRegex◬` the collision chance is really low so we can get away with just calling `replace` on it.

Reply:
yeah, I'd agree. that might simplify the code significantly.

Reply:
I suppose, my gut reaction is that I don't like it because it's not as "correct". Though practically it's very unlikely the string will have some obscure character like that, and even if it does it just means the debug string will be slightly off, so I suppose I can do that